From 6aa408a4e365488d2bf2516515921ef502a887c4 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:20:12 -0400
Subject: [fs] nfsd: Ensure that the laundromat unhashes the client before releasing locks

Message-id: <1410898849-22126-129-git-send-email-bfields@redhat.com>
Patchwork-id: 95847
O-Subject: [Patch RHEL7 128/165] nfsd: Ensure that the laundromat unhashes the client before releasing locks
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

If we leave the client on the confirmed/unconfirmed tables, and leave
the sessions visible on the sessionid_hashtbl, then someone might
find them before we've had a chance to destroy them.

Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 4864af97e02d1ef6aa78963195a64ed2ed7752c3
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index cb630db..a374592 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -4125,13 +4125,15 @@ nfs4_laundromat(struct nfsd_net *nn)
     clp->cl_clientid.cl_id);
    continue;
   }
-  list_move(&clp->cl_lru, &reaplist);
+  unhash_client_locked(clp);
+  list_add(&clp->cl_lru, &reaplist);
  }
  spin_unlock(&nn->client_lock);
  list_for_each_safe(pos, next, &reaplist) {
   clp = list_entry(pos, struct nfs4_client, cl_lru);
   dprintk("NFSD: purging unused client (clientid %08x)\n",
    clp->cl_clientid.cl_id);
+  list_del_init(&clp->cl_lru);
   expire_client(clp);
  }
  spin_lock(&state_lock);
-- 
1.7.1